### PR TITLE
[for master] tpm2_create: Use better object attributes defaults for authentication

### PR DESCRIPTION
**Pull request by martinezjavier**
The tpm2_create tool allows to define a policy session or a password for
authentication. By default no policy session is used and the password is
empty, which means that this empty password is used for authentication.

So the default object attribute flag userWithAuth is set in order to use
the empty password. This isn't a good default though if a policy is set,
since in this case the policy session has to be used for authentication
instead of an empty password.

If a policy is defined, the userWithAuth bit has to be clear unless the
user defines a password so in that case authentication would happen only
using the policy session or the defined password.

Also add these cases in the integration test to detect regressions.

Fixes: #1123

Signed-off-by: Javier Martinez Canillas <javierm@redhat.com>

### CODE REVIEW

**Comment by diabonas** [comment]
I can confirm that this works as expected from a user perspective: if a policy is specified, but no password, I can only unseal the secret using a policy. If both a password and a policy are specified, I can unseal using either one. If no policy is specified, I can unseal with the given password (or the empty password if none is specified).

I noticed that if a policy is specified, `adminwithpolicy` is always set, even if a password is set as well. This means that operations requiring the ADMIN role can only be done using the policy, but not with the password. Is this intended? I am not actually sure what operations need an ADMIN role authorization, but a quick glance at the TPM specification shows that this apparently affects TPM2_Certify, TPM2_ActivateCredential and TPM2_NV_ChangeAuth (so probably not too interesting for sealed secrets). Anyway, unsealing is not affected by `adminwithpolicy`, as this is a USER role operation.

**Comment by martinezjavier** [comment]
@diabonas yes, I also noticed that but since is the behavior when both `userWithAuth` and `adminWithPolicy` bits are set, I thought that it was expected.

**Comment by martinezjavier** [comment]
I see that the `unseal.sh` integration test is failing, but it fails for me locally even without my changes:

```
$ TPM2_DEVICE=/dev/tpmrm0 ./tests/unseal.sh

--- To recreate this test run the following from: /home/javier/devel/tpm2-tools/test/integration ---
#!/usr/bin/env bash
export TPM2_ABRMD="" TPM2_SIM="" TPM2_DEVICE="/dev/tpmrm0" PATH="/home/javier/perl5/bin:/home/javier/perl5/bin:/usr/libexec/python2-sphinx:/usr/lib64/qt-3.3/bin:/usr/share/Modules/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/javier/.local/bin:/home/javier/bin:/home/javier/bin:/home/javier/devel/u-boot/tools/patman" TPM2_SIM_NV_CHIP="" TPM2_TOOLS_TEST_FIXTURES=""
./tests/unseal.sh
--- EOF ---

creating simulator working dir: /tmp/tpm2_test_qkLV10
/tmp/tpm2_test_qkLV10 ~/devel/tpm2-tools/test/integration
Switched to CWD: /tmp/tpm2_test_qkLV10
Running tpm2_clear
ERROR: Tss2_Sys_CreatePrimary(0x2C4) - tpm:parameter(2):value is out of range or is not correct for the context
ERROR: Unable to run tpm2_createprimary
tpm2_createprimary -Q -a e -g $alg_primary_obj -G $alg_primary_key -o $file_primary_key_ctx on line 74 failed: 1
Shutting down
~/devel/tpm2-tools/test/integration
Removing sim dir: /tmp/tpm2_test_qkLV10
```

**Comment by diabonas** [comment]
@martinezjavier As far as I understand it, `userwithauth` and `adminwithpolicy` have an "inverse" logic: the TPM specification says

> This attribute is similar to *userWithAuth* but the logic is a bit different. When *userWithAuth* is CLEAR, the *authValue* may not be used for USER mode authorizations. When *adminWithPolicy* is CLEAR, it means that the *authValue* may be used for ADMIN role. Policy may always be used regardless of the setting of *userWithAuth* or *adminWithPolicy*.

So I think if `userwithauth` is set, we probably want to have `adminwithpolicy` unset in general (and the other way round, if `adminwithpolicy` is set, `userwithauth` should be unset) so that USER and ADMIN role use the same authorization. I am not too sure on this subject though, so I do not have a strong opinion here (it works fine for unsealing either way).

Anyway, I just realised I commented on the wrong pull request: I tested the one for the `3.X` branch, which worked fine. This pull request for the `master` branch currently does not work as expected for me, apparently because the attribute setting in `on_option()` is overridden by the default attribute specification [in l. 233](https://github.com/martinezjavier/tpm2-tools/blob/8f8a60a30e1095381216f6852e3f24713839c9e2/tools/tpm2_create.c#L233). This might also be the reason for the failed test.

**Comment by diabonas** [comment]
@martinezjavier Never mind, the error you get is due to https://github.com/tpm2-software/tpm2-tools/issues/1079 - changing [l. 37 in `unseal.sh`](https://github.com/tpm2-software/tpm2-tools/blob/0e37ed6dd329fa87a9505631fa5ac841e2768d40/test/integration/tests/unseal.sh#L37) to `alg_primary_key=ecc:null` until that bug is sorted out should fix it.

**Comment by diabonas** [comment]
> But then we should probably error if a user tries to set both `--pwdk` and `--policy-file` since are mutually exclusive? Otherwise setting a password will clear `adminwithpolicy` which will be confusing for users again. For me is confusing that both could be used for USER role but the password can only be used for ADMIN when there's no policy for the session.

Technically `--pwdk` and `--policy-file` are not exclusive, they just might not do what the user think they do: instead of requiring password *and* policy to unseal the data, you can use either one of those if `userwithauth` is set.

I agree that the same authentication should be used for USER and ADMIN role: so the attributes should be
- no password, no policy: `userwithauth` (authentication with empty password)
- no password, but a policy: `adminwithpolicy` (so that the ADMIN role cannot be assumed with the empty password)
- password, but no policy: `userwithauth` (no `adminwithpolicy`, otherwise ADMIN role could not be assumed at all due to lack of policy)
- password and policy: `userwithauth` (no `adminwithpolicy`, otherwise USER role could be assumed with password, but ADMIN role only with policy).

Currently we set `adminwithpolicy` also in the last case, while the other cases work as expected.

> No, the default is only set if no attribute was in `on_option(). Look at: https://github.com/martinezjavier/tpm2-tools/blob/8f8a60a30e1095381216f6852e3f24713839c9e2/lib/tpm2_alg_util.c#L939

You are right. Then there must be some other problem, because this is what I get with the version from this pull request:

```
# tpm2_createprimary --quiet --halg sha256 --kalg rsa2048:null:aes128cfb --out-context primary.ctx
# tpm2_createpolicy --policy-pcr --set-list sha256:0,1,2,3 --policy-file key.pol
# echo "secret" | tpm2_create --context-parent primary.ctx --policy-file key.pol --in-file -
name-alg:
  value: sha256
  raw: 0xb
attributes:
  value: fixedtpm|fixedparent|userwithauth
  raw: 0x52
type:
  value: keyedhash
  raw: 0x8
algorithm: 
  value: null
  raw: 0x10
keyedhash: 4268870ed648fd20a91a77c064da9d6e6dcf79616bde8f36ed0d95e1d80e76a1
authorization policy: 2388d83a3df40e8f53b8e201fc24ea9a996a107996409306e783eb10d89abe98
```

As you can see, `userwithauth` is still set, which should not be the case if a policy and no password is specified.

> I see, but I prefer that to be a different commit so I'll just leave the commit as it is now.

I agree, this should hopefully be fixed in https://github.com/tpm2-software/tpm2-tools/issues/1079 anyway, making this change unnecessary.

**Comment by martinezjavier** [comment]
> password and policy: userwithauth (no adminwithpolicy, otherwise USER role could be assumed with password, but ADMIN role only with policy).
> Currently we set adminwithpolicy also in the last case, while the other cases work as expected.

Right, but if we clear `adminwithpolicy` in this case then the policy will never be used. That's why I said that we should probably error if a user tries to set both. Or am I misunderstanding?

> As you can see, userwithauth is still set, which should not be the case if a policy and no password is specified.

That's weird... unfortunately I'm not able to test the patch since `tpm2_createprimary` fails for me as mentioned in PR #1127.

**Comment by martinezjavier** [comment]
> That's weird... unfortunately I'm not able to test the patch since tpm2_createprimary fails for me as mentioned in PR #1127.

it fails for me even after setting `alg_primary_key=ecc:null` as suggested:

```
diff --git a/test/integration/tests/unseal.sh b/test/integration/tests/unseal.sh
index a835dac52b2..52eb534d2bf 100755
--- a/test/integration/tests/unseal.sh
+++ b/test/integration/tests/unseal.sh
@@ -34,7 +34,7 @@
 source helpers.sh
 
 alg_primary_obj=sha256
-alg_primary_key=ecc
+alg_primary_key=ecc:null
 alg_create_obj=sha256
 alg_pcr_policy=sha1
```

**Comment by martinezjavier** [comment]
Ah, you edited your original comment and instead I should use `alg_primary_key=rsa2048:null:aes128cfb`.

**Comment by diabonas** [comment]
> Right, but if we clear `adminwithpolicy` in this case then the policy will never be used. That's why I said that we should probably error if a user tries to set both. Or am I misunderstanding?

Well, it could still be used for both USER and ADMIN role as an alternative the password. The use case would probably be something like this: you want to seal a secret to a certain PCR configuration, but retain a way to unseal the secret even if the configuration changes. To do so, you create a secret with `--pwdk` and `--policy-file`, store the password in a safe place and use the policy for normal operations. If the PCR configuration changes, you are still able to access the secret using the backup password. In this case, you probably also want to be able to do admin tasks with the password, as you are otherwise completely locked out of the ADMIN role if you cannot reproduce the PCR configuration.

>Ah, you edited your original comment and instead I should use `alg_primary_key=rsa2048:null:aes128cfb`.
Yeah, I got the parameter wrong in the first attempt, sorry. This should hopefully work better, if not, I am happy to try to find the cause of the problem myself later this evening.

**Comment by martinezjavier** [comment]
> Yeah, I got the parameter wrong in the first attempt, sorry. This should hopefully work better, if not, I am happy to try to find the cause of the problem myself later this evening.

Great, I've pushed a new branch that works now (modulo the change for `alg_primary_key` of course).

**Comment by martinezjavier** [comment]
> Well, it could still be used for both USER and ADMIN role as an alternative the password. The use case would probably be something like this: you want to seal a secret to a certain PCR configuration, but retain a way to unseal the secret even if the configuration changes. To do so, you create a secret with --pwdk and --policy-file, store the password in a safe place and use the policy for normal operations. If the PCR configuration changes, you are still able to access the secret using the backup password. In this case, you probably also want to be able to do admin tasks with the password, as you are otherwise completely locked out of the ADMIN role if you cannot reproduce the PCR configuration.

Yes, but in this case you *won't* be able to use the password for ADMIN role. In others words would either:

a) Can use either password or policy session for USER or
b) Can only use password for both USER or ADMIN.

But you can't have either for USER and have password for ADMIN.

**Comment by martinezjavier** [comment]
@diabonas what I tried to say is that we only have one `userWithAuth` and `adminWithPolicy` but with different semantics for USER and ADMIN role.

**Comment by diabonas** [comment]
> Great, I've pushed a new branch that works now (modulo the change for `alg_primary_key` of course).

Nice, that works as expected for the `master` branch (the same way it does on the `3.X` branch), thank you very much!

> Yes, but in this case you won't be able to use the password for ADMIN role. In others words would either:
> 
> a) Can use either password or policy session for USER or
> b) Can only use password for both USER or ADMIN.
> 
> But you can't have either for USER and have password for ADMIN.

You cannot disable authentication with a policy for either USER or ADMIN, this is always allows regardless of the attributes: the specification says

> 1\) ***userWithAuth*** – If this attribute is SET, then USER role authorization for an object may be provided with an HMAC session or a password. If this attribute is CLEAR, then the *authValue* may not be used for USER role authorization, meaning that authorization may not be done using an HMAC session or a password. USER role authorizations with a policy are always allowed regardless of the setting of this attribute.
> 2) ***adminWithPolicy*** – If this attribute is SET, then ADMIN role authorization for an object may only be provided with a policy session. If this attribute is CLEAR, then authorization may be provided with a policy session, with an HMAC session, or with a password.

So `userwithauth` and `adminwithpolicy` only control whether you can *additionally* use the provided password as an authorization: while a set `userwithauth` means you *can* use a password for the USER role, a set *adminwithpolicy* set means you *cannot* use a password for the ADMIN role. I maintain that if you set `userwithauth` to allow passwords for the USER role, you should clear `adminwithpolicy` so as to allow passwords for the ADMIN role as well.

| Attribute | USER with password | USER with policy | ADMIN with password | ADMIN wih policy |
|------------------------------|--------------------|------------------|---------------------|------------------|
| \<none> | no | yes | yes | yes |
| userwithauth | yes | yes | yes | yes |
| adminwithpolicy | no | yes | no | yes |
| userwithauth\|adminwithpolicy | yes | yes | no | yes |

The first and last row are asymmetric with regards to the allowed authorization for USER and ADMIN, so I think usually you should use exactly one of the two attributes: either only `userwithauth` to allow passwords for both roles (in addition to policy sessions, which are always allowed), or only `adminwithpoliy` to only allow policy sessions for both.

**Comment by martinezjavier** [comment]
> You cannot disable authentication with a policy for either USER or ADMIN, this is always allows regardless of the attributes: the specification says

Ah, I misunderstood the options then. I've read so many times the spec and I keep forgetting the details... maybe I'm just stupid.



> The first and last row are asymmetric with regards to the allowed authorization for USER and ADMIN, so I think usually you should use exactly one of the two attributes: either only userwithauth to allow passwords for both roles (in addition to policy sessions, which are always allowed), or only adminwithpoliy to only allow policy sessions for both.

But then why would we want to set `adminWithpolicy` at all? Why not just set `userWithAuth` if either a password is provided or is an empty password (default) and no policy provided?

**Comment by martinezjavier** [comment]
> I think in the case of no policy or no password, we should error out and inform the user to do one of those two things.

I'm not sure about this. I think there's a difference between:

a) I'm providing a PCR policy (so I expect the data to be sealed against a specific PCR state) and
b) I'm not providing neither a PCR policy not a password (so I'm binding a secret against a specific TPM, I don't care about the PCR state).

Both are valid use cases, the problem is that we were breaking an assumption for (a). But if we error if the user didn't provide neither a password not a PCR policy, then we will break (b).

File: tools/tpm2_create.c [hunk 1]
```diff
@@ -261,6 +261,12 @@ int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
         goto out;
     }
 
+    if (ctx.flags.L) {
+        if (!ctx.flags.p)
+             ctx.in_public.publicArea.objectAttributes &= ~TPMA_OBJECT_USERWITHAUTH;
+        ctx.in_public.publicArea.objectAttributes |= TPMA_OBJECT_ADMINWITHPOLICY;
```
[the remark this answers is not in the document]
Comment (martinezjavier) [reply tools/tpm2_create.c]: @williamcroberts sorry about the brackets, I haven't worked on this codebase for a while and forgot the coding style...

Yeah, it's a pity that we didn't have that behavior for 3.X, but for now I would just have slightly different patches for 3.X and master.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Javier Martinez Canillas [mailto:notifications@github.com]
> Sent: Monday, July 30, 2018 8:43 AM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Comment
> <comment@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-tools] [for master] tpm2_create: Use better
> object attributes defaults for authentication (#1127)
> 
> 	I think in the case of no policy or no password, we should error out and
> inform the user to do one of those two things.
> 
> I'm not sure about this. I think there's a difference between:
> 
> a) I'm providing a PCR policy (so I expect the data to be sealed against a specific
> PCR state) and
> b) I'm not providing neither a PCR policy not a password (so I'm binding a secret
> against a specific TPM, I don't care about the PCR state).

Totally valid, but indicate somehow that you didn't want any protections on it to
the tool via  -P""

> 
> Both are valid use cases, the problem is that we were breaking an assumption for
> (a). But if we error if the user didn't provide neither a password not a PCR policy,
> then we will break (b).

We error if they never specified -P, we don't enforce it to have a length.

**Comment by martinezjavier** [comment]
> Totally valid, but indicate somehow that you didn't want any protections on it to
> the tool via  -P""

Agreed for master, but I don't think that's suitable for 3.X since that would be another user visible change in `tpm2_create`.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Javier Martinez Canillas [mailto:notifications@github.com]
> Sent: Monday, July 30, 2018 8:48 AM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-tools] [for master] tpm2_create: Use better
> object attributes defaults for authentication (#1127)
> 
> 	Totally valid, but indicate somehow that you didn't want any protections
> on it to
> 	the tool via -P""
> 
> Agreed for master, but I don't think that's suitable for 3.X since that would be
> another user visible change in tpm2_create.

Oh yeah definitely. I may have crossed wires on my comments between 3.X and master,
but I meant for master.

**Comment by martinezjavier** [comment]
@williamcroberts ah cool. Is just that I saw the same comment in both PR (probably my bad for opening too many PR...).

So I'll re-spin the 3.X PR with @diabonas and your suggestion for now, once we agreed on what goes there then we can move to master. Thanks for your feedback!

**Comment by diabonas** [comment]
@martinezjavier 
> Ah, I misunderstood the options then. I've read so many times the spec and I keep forgetting the details... maybe I'm just stupid.

Nah, I blame the specification for defining two attributes  that essentially do the same thing with exactly opposite semantics ;)

> But then why would we want to set `adminWithpolicy` at all? Why not just set `userWithAuth` if either a password is provided or is an empty password (default) and no policy provided?

Fair point! My fear (not backed by any evidence at all, I just do not know the specs well enough to exclude the possibility) would be that you can do something like changing the authorization policy of a key to a different (attacker-controller) PCR policy with the ADMIN role - this would obviously be bad if you can just assume that role with the default empty password. Another glance over the specification  leads me to believe however that you can do nothing interesting with the ADMIN role for keys, so leaving `adminwithpoliy` completely alone is probably fine.

I see now that @williamcroberts  has independently suggested not setting `adminwithpolicy` in  https://github.com/tpm2-software/tpm2-tools/issues/1097#issuecomment-408898320 as well, so I think this probably the best way to go forward to not overcomplicate matters.

**Comment by diabonas** [comment]
@martinezjavier Now that https://github.com/tpm2-software/tpm2-tools/pull/1167 has been merged, I think we can have the improved setting of the `userwithauth` object attribute on the master branch as well: until now, this was blocked by having to manually change the primary key algorithm for some hardware TPMs as described in https://github.com/tpm2-software/tpm2-tools/pull/1126#issuecomment-408930034, which should not be necessary any more.

It seems that the changes in this pull request have been overwritten (by https://github.com/tpm2-software/tpm2-tools/pull/1136), therefore https://github.com/tpm2-software/tpm2-tools/pull/1126 needs to be cherry-picked for the current master branch.

**Comment by diabonas** [comment]
@martinezjavier I tried to reconstruct the last good state of this pull request from the [pull request for the 3.X branch](https://github.com/tpm2-software/tpm2-tools/pull/1126) and the preserved diffs in this thread. I think it looked like the attached patch, which I confirm to work as described in https://github.com/tpm2-software/tpm2-tools/pull/1126#issuecomment-408918917, with the integration test passing against the simulator. I do not feel it appropriate to create a new pull request for this though, as it is still essentially your code :)
```
diff --git a/test/integration/tests/unseal.sh b/test/integration/tests/unseal.sh
index 4db5a5c..4ad6a16 100755
--- a/test/integration/tests/unseal.sh
+++ b/test/integration/tests/unseal.sh
@@ -110,4 +110,54 @@ unsealed=`tpm2_unseal -c $file_unseal_key_ctx -L ${alg_pcr_policy}:${pcr_ids} -F
 
 test "$unsealed" == "$secret"
 
+# Test that unseal fails if a PCR policy isn't provided
+
+trap - ERR
+
+tpm2_unseal -c $file_unseal_key_ctx 2> /dev/null
+if [ $? != 1 ]; then
+  echo "tpm2_unseal didn't fail without a PCR policy!"
+  exit 1
+fi
+
+# Test that unseal fails if PCR state isn't the same as the defined PCR policy
+
+pcr_extend=$(echo $pcr_ids | cut -d ',' -f1)
+
+tpm2_pcrextend $pcr_extend:sha1=6c10289a8da7f774cf67bd2fc8502cd4b585346a
+
+tpm2_unseal -c $file_unseal_key_ctx -L ${alg_pcr_policy}:${pcr_ids} -F $file_pcr_value 2> /dev/null
+if [ $? != 1 ]; then
+  echo "tpm2_unseal didn't fail with a PCR state different than the policy!"
+  exit 1
+fi
+
+# Test that the object can be unsealed without a policy but a password
+
+trap onerror ERR
+
+rm $file_unseal_key_pub $file_unseal_key_priv $file_unseal_key_name
+
+tpm2_pcrlist -Q -L ${alg_pcr_policy}:${pcr_ids} -o $file_pcr_value
+
+tpm2_createpolicy -Q -P -L ${alg_pcr_policy}:${pcr_ids} -F $file_pcr_value -f $file_policy
+
+tpm2_create -Q -g $alg_create_obj -u $file_unseal_key_pub -r $file_unseal_key_priv -I- -C $file_primary_key_ctx -L $file_policy -p secretpass <<< $secret
+
+tpm2_load -Q -C $file_primary_key_ctx  -u $file_unseal_key_pub  -r $file_unseal_key_priv -n $file_unseal_key_name -o $file_unseal_key_ctx
+
+unsealed=`tpm2_unseal -c $file_unseal_key_ctx -p secretpass`
+
+test "$unsealed" == "$secret"
+
+# Test that unseal fails when using a wrong password
+
+trap - ERR
+
+tpm2_unseal -c $file_unseal_key_ctx -p wrongpass 2> /dev/null
+if [ $? != 1 ]; then
+  echo "tpm2_unseal didn't fail when using a wrong object password!"
+  exit 1
+fi
+
 exit 0
diff --git a/tools/tpm2_create.c b/tools/tpm2_create.c
index 7b9fdb4..4ce3778 100644
--- a/tools/tpm2_create.c
+++ b/tools/tpm2_create.c
@@ -261,6 +261,10 @@ int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
         goto out;
     }
 
+    if (!ctx.flags.p) {
+        ctx.in_public.publicArea.objectAttributes &= ~TPMA_OBJECT_USERWITHAUTH;
+    }
+
     if (ctx.flags.I && ctx.in_public.publicArea.type != TPM2_ALG_KEYEDHASH) {
         LOG_ERR("Only TPM2_ALG_KEYEDHASH algorithm is allowed when sealing data");
         goto out;
```

**Pull request closed without merging**